backends/ebpf/runtime/ebpf_runtime.c

```diff
@@ -14,89 +14,192 @@ See the License for the specific language governing permissions and
+ * @details This is a mock function emulating the behavior of a running
+ * eBPF program. It takes a list of input packets and iteratively parses them
+ * using the given imported ebpf_filter function. The output defines whether
+ * or not the packet is "dropped." If the packet is not dropped, its content is
+ * copied and appended to an output packet list.
+ *
+ * @param pkt_list A list of input packets running through the filter.
+ * @return The list of packets "surviving" the filter function
+ */
+static pcap_list_t *feed_packets(pcap_list_t *pkt_list) {
+    pcap_list_t *output_pkts = allocate_pkt_list();
+    uint64_t list_len = get_pkt_list_length(pkt_list);
+    for (uint64_t i = 0; i < list_len; i++) {
+        /* Parse each packet in the list and check the result */
         struct sk_buff skb;
-        skb.data = (void *) packet;
-        skb.len = pcap_hdr->len;
+        pcap_pkt *input_pkt = get_packet(pkt_list, i);
+        skb.data = (void *) input_pkt->data;
+        skb.len = input_pkt->pcap_hdr.len;
         int result = ebpf_filter(&skb);
-        if (result)
-            pcap_dump((unsigned char *) out_handle, pcap_hdr, packet);
+        if (result != 0) {
+            /* We copy the entire content to emulate an outgoing packet */
+            pcap_pkt *out_pkt = copy_pkt(input_pkt);
+            output_pkts = append_packet(output_pkts, out_pkt);
+        }
+        if (debug)
+            printf("Result of the eBPF parsing is: %d\n", result);
+    }
+    return output_pkts;
+}
+
+static void write_pkts_to_pcaps(const char *pcap_base, pcap_list_array_t *output_array) {
+    uint16_t arr_len = get_list_array_length(output_array);
+    for (uint16_t i = 0; i < arr_len; i++) {
+        char pcap_out_name[FILE_NAME_MAX];
+        if (snprintf(pcap_out_name, FILE_NAME_MAX, "%s%d%s", pcap_base, i, PCAPOUT) >=FILE_NAME_MAX) {
+            fprintf(stderr, "Filename %s%d%s too long. Truncated.\n"
+                ,pcap_base, i, PCAPIN);
+            exit(EXIT_FAILURE);
+        }
+        if (debug)
+            printf("Processing output file: %s\n", pcap_out_name);
+        pcap_list_t *out_list = get_list(output_array, i);
+        write_pkts_to_pcap(pcap_out_name, out_list);
+    }
+}
+
+static void *run_and_record_output(const char *pcap_base, pcap_list_t *pkt_list) {
+    /* Create an array of packet lists */
+    pcap_list_array_t *output_array = NULL;
+    /* Feed the packets into our "loaded" program */
+    pcap_list_t *output_pkts = feed_packets(pkt_list);
+    /* Split the output packet list by interface. This destroys the list. */
+    output_array = delete_and_split_list(output_pkts, output_array);
+    /* Write each list to a separate pcap output file */
+    write_pkts_to_pcaps(pcap_base, output_array);
+    /* Delete the array, including the data it is holding */
+    delete_array(output_array);
+}
+
+static pcap_list_t *get_packets(const char *pcap_base, uint16_t num_pcaps, pcap_list_t *merged_list) {
```

Comment:
If this function deletes its input argument merged_list the name should also reflect that fact.
In general people don't expect functions to deallocate their arguments.